src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -125,6 +125,14 @@ private BoundStatement RewriteEnumeratorForEachStatement(BoundForEachStatement n
             //     V v = (V)(T)e.Current;
             //     /* node.Body */
             // }
+
+            // The scope of the iteration variable is the foreach statement syntax.
```

Comment:
First of all, for EnC mapping to work no two BoundBlocks that represent a closure in the lambda rewriter can be associated with the same syntax node. If they were the mapping would be broken.
You are correct that once expression-level declarations are introduced we need to amend the implementation of IsClosureScope (and of other code in EnC infrastructure). And that is true regardless of what node this particular BoundBlock is associated with. 
For now we have two options:
1) The small change implemented in this PR. 
2) A bigger change of updating the code-gen to not emit a distinct display class and instead share the closure with variables of the embedded statement.
I'm not against implementing [2] for RTM. I believe [1] is better for RC, since it's a less risky change.
BTW, the node associated with the BoundBlock does not need to exactly describe the scope of the variable. It does however need to uniquely represent it, meaning that it should be possible to unambiguously derive the exact span from the associated node. Given scope associated with a node of type ForEachStatementSyntax we know that the exact scope being represented spans the embedded statement.